src/data/transcript.rs

```diff
@@ -0,0 +1,51 @@
+use core::result::Result;
+use std::path::PathBuf;
+
+use crate::SharedTranscript;
+use serde::{de::DeserializeOwned, ser::Serialize};
+
+pub trait Contribution: Serialize + DeserializeOwned {
+    type Receipt: Serialize;
+    fn get_receipt(&self) -> Self::Receipt;
+}
+
+pub trait Transcript: Serialize + DeserializeOwned {
+    type ContributionType: Contribution;
+    type ValidationError: Serialize;
+
+    fn verify_contribution(
+        &self,
+        contribution: &Self::ContributionType,
+    ) -> Result<(), Self::ValidationError>;
+
+    fn update(&self, contribution: &Self::ContributionType) -> Self;
+
+    fn get_contribution(&self) -> Self::ContributionType;
+}
+
+pub async fn read_transcript_file<T: DeserializeOwned + Send + 'static>(path: PathBuf) -> T {
+    let handle = tokio::task::spawn_blocking::<_, T>(|| {
+        let f = std::fs::File::open(path).expect("can't access transcript file.");
+        let reader = std::io::BufReader::new(f);
+        serde_json::from_reader::<_, T>(reader).expect("unreadable transcript")
+    });
+    handle.await.expect("can't read transcript")
+}
+
+pub async fn write_transcript_file<T: Transcript + Send + Sync + 'static>(
+    target_path: PathBuf,
+    work_path: PathBuf,
+    transcript: SharedTranscript<T>,
+) {
+    let handle = tokio::task::spawn_blocking(move || {
+        let f = std::fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .open(&work_path)
+            .expect("Can't access transcript file.");
+        let transcript = transcript.blocking_read();
+        serde_json::to_writer_pretty(&f, &*transcript).expect("Cannot write transcript");
+        std::fs::rename(&work_path, &target_path).unwrap();
```

Comment:
`expect(` and `unwrap(` should be avoided as much as possible and instead we should bubble `Result` up. The `eyre::Result` makes this easy because it's a catch-all type.
We want this so we get proper error reporting and correct shutdown.

I only ever use `expect(` when I know that the operation can trivially never fail (like getting the first item from a list initialized in the same code block) and adding a Result type would hurt the devex. I never use unwrap in prod code.